File: agent/src/life_update_agent/inference/vision_models.py
"""Curated engine choices for turning a screenshot into text/description.

Mirrors `inference/models.py`'s pattern exactly. "tesseract" is a special
case, not an Ollama tag: it's a bundled local binary, runs inline in the
capture loop, and never needs pulling. The other two are Ollama vision
models, pulled the same way session-summarization models are.

qwen2.5vl was chosen over moondream/llava for the vision options: moondream
is explicitly weak on dense text, and our content (code, docs, terminals)
is exactly dense text - qwen2.5vl is specifically strong at structured,
text-heavy screenshots.
"""

from __future__ import annotations

from dataclasses import dataclass

DEFAULT_VISION_ENGINE = "tesseract"

TESSERACT_ENGINE = "tesseract"


@dataclass(frozen=True)
class VisionChoice:
    name: str
    size_human: str
    description: str


VISION_CHOICES: list[VisionChoice] = [
    VisionChoice(TESSERACT_ENGINE, "~35 MB", "fast, text-only, runs inline (recommended default)"),
    VisionChoice("qwen2.5vl:3b", "3.2 GB", "reads screen content semantically, runs when idle"),
    VisionChoice("qwen2.5vl:7b", "6.0 GB", "higher quality, slower"),
]

VISION_ENGINE_NAMES = {v.name for v in VISION_CHOICES}


def find_vision_engine(name: str) -> VisionChoice | None:
    return next((v for v in VISION_CHOICES if v.name == name), None)


def is_ollama_backed(name: str) -> bool:
    return name != TESSERACT_ENGINE

File: app-shell/src-tauri/src/vision_models.rs
//! Curated vision-engine registry for screen watching. "native" is Apple's
//! Vision framework OCR - on-device, hardware-accelerated, nothing bundled
//! (it replaced the old Tesseract dependency). The qwen options are Ollama
//! vision models for semantic descriptions, via the user's own Ollama.

use serde::Serialize;

#[derive(Serialize, Clone, Copy)]
pub struct VisionChoice {
    pub name: &'static str,
    pub size_human: &'static str,
    pub description: &'static str,
}

pub const NATIVE_ENGINE: &str = "native";
pub const DEFAULT_VISION_ENGINE: &str = NATIVE_ENGINE;

pub const VISION_CHOICES: &[VisionChoice] = &[
    VisionChoice {
        name: NATIVE_ENGINE,
        size_human: "built-in",
        description: "Apple Vision OCR - on-device, instant, reads screen text (recommended)",
    },
    VisionChoice { name: "qwen2.5vl:3b", size_human: "3.2 GB", description: "describes screen content semantically, runs when idle (via the Ollama app)" },
    VisionChoice { name: "qwen2.5vl:7b", size_human: "6.0 GB", description: "higher quality, slower (via the Ollama app)" },
];

pub fn is_ollama_backed(name: &str) -> bool {
    // "tesseract" is the pre-rewrite name for the non-Ollama path; configs
    // written by older builds may still carry it.
    name != NATIVE_ENGINE && name != "tesseract"
}
